From 435a801c10a50ceb14bb434a54b7c5bcfd0b2dcc Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 18 Aug 2014 19:58:50 -0400
Subject: [block] nvme: Use last bytes of f/w rev SCSI Inquiry

Message-id: <1408391935-24886-18-git-send-email-dmilburn@redhat.com>
Patchwork-id: 87890
O-Subject: [RHEL7.1 PATCH BZ 1111259 17/22] NVMe: Use last bytes of f/w rev SCSI Inquiry
Bugzilla: 1111259
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Keith Busch <keith.busch@intel.com>

After skipping right-padded spaces, use the last four bytes of the
firmware revision when reporting the Inquiry Product Revision. These
are generally more indicative to what is running.

Signed-off-by: Keith Busch <keith.busch@intel.com>
Acked-by: Vishal Verma <vishal.l.verma@linux.intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit dedf4b156134e0dedec18ebecda3e74077fa7c92)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-scsi.c b/drivers/block/nvme-scsi.c
index 13403ee..4fc25b9 100644
--- a/drivers/block/nvme-scsi.c
+++ b/drivers/block/nvme-scsi.c
@@ -681,6 +681,7 @@ static int nvme_trans_standard_inquiry_page(struct nvme_ns *ns,
  u8 resp_data_format = 0x02;
  u8 protect;
  u8 cmdque = 0x01 << 1;
+ u8 fw_offset = sizeof(dev->firmware_rev);
 
  mem = dma_alloc_coherent(&dev->pci_dev->dev, sizeof(struct nvme_id_ns),
     &dma_addr, GFP_KERNEL);
@@ -716,7 +717,11 @@ static int nvme_trans_standard_inquiry_page(struct nvme_ns *ns,
  inq_response[7] = cmdque; /* wbus16=0 | sync=0 | vs=0 */
  strncpy(&inq_response[8], "NVMe    ", 8);
  strncpy(&inq_response[16], dev->model, 16);
- strncpy(&inq_response[32], dev->firmware_rev, 4);
+
+ while (dev->firmware_rev[fw_offset - 1] == ' ' && fw_offset > 4)
+  fw_offset--;
+ fw_offset -= 4;
+ strncpy(&inq_response[32], dev->firmware_rev + fw_offset, 4);
 
  xfer_len = min(alloc_len, STANDARD_INQUIRY_LENGTH);
  res = nvme_trans_copy_to_user(hdr, inq_response, xfer_len);
-- 
1.7.1